doc/dev/kdb-operations.md

```diff
@@ -0,0 +1,156 @@
+- `proc:/` mountpoints may exist, but they are read-only (see below)
+- `dir:/`, `user:/` and `system:/` mountpoints can be created without restrictions
+
+Other restrictions:
+
+- Creating mountpoints for root keys (`system:/`, `user:/`, etc.) is not allowed.
+  These parts of the KDB always use hardcoded backends that can only be configured at compile-time.
+- Creating a mountpoint for `system:/elektra` or any mountpoint below is not allowed.
+  This part of the KDB will always be stored in the bootstrap backend.
+  While the only part that is actually required for the bootstrap process is the one below `system:/elektra/mountpoints`, everything below `system:/elektra` is reserved and should only be used internally.
+  Additionally, `system:/elektra/version` and `system:/elektra/modules` will always point to hardcoded read-only backends containing information about this Elektra installation.
+
+## `get` Operation
+
+The purpose of the `get` operation is to read data stored in backends into a `KDB` instance.
+
+> **Note:** Some details of a `get` operation are defined in the [contract with backend plugins](./backend-plugins.md).
+
+Properties of `kdbGet()`:
+
+- After calling `kdbGet (kdb, ks, parentKey)`, the KeySet `ks` will contain _all keys_ (including their values) that are stored in _any backend_ with a mountpoint that is _below `parentKey`_.
+- After calling `kdbGet (kdb, ks, parentKey)`, below `parentKey` the KeySet `ks` will _only_ contain keys that are stored in a backend.
+- The KeySet `ks` may contain other keys not below `parentKey`.
+  These keys fall into one of two categories:
+  1. Keys that are stored in a backend that is not below `parentKey`.
+     `kdbGet()` may decide that it is more efficient (e.g. because of a cache) to return more keys than requested.
+  2. Keys that were already present in `ks` when `kdbGet()` was called.
+
+In simpler terms, this means to the caller it looks as if `kdbGet()` had removed all keys below `parentKey` from `ks` and then loaded the data from the backends.
+Below `parentKey` the KeySet `ks` correctly represents the state of the KDB.
+For the rest of `ks` there are no such guarantees.
+
+TODO: cache correct? notifications? (just call at the end?)
+
+The basic flow of this operation is:
+
+1. Determine the backends needed to read all keys below `parentKey`
+2. Run the `init` phase on all the backends that haven't been initialized
+3. Run the `resolver` phase on all backends
+4. From now on ignore all backends, which indicated that there is no update.
+5. If all backends are now ignored, **return**.
+6. If a global cache plugin is enabled:
+   Ask the global cache plugin for the cache times for all backends.
+7. If all backends have an existing cache entry:
+   Run the `cachecheck` phase on all backends
+8. If all backends indicated the cache is still valid:
+   Ask the global cache plugin for the cached data and **return**.
```

Comment:
Yes, this does some checks and if everything is fine loads the cached Split struct as well as the cached KeySet.

I think removing the split also implies changes in mmapstorage. This PR changes some firm assumptions that were made for the cache. I did not expect the Split going away, neither did I expect the whole backend stuff to be changed.

Reply:
>  I assume storing a second KeySet in the cache shouldn't be too hard (at least not harder than storing the Split).

Actually, the split data is already stored within a keyset for the cache. Since I already wrote the serialization code for keysets I tried to squeeze everything into a keyset.

Reply:
From what I can tell, the `cache` plugin doesn't know about the `Split` structure anyway. Instead the `Split` is written into the global keyset, which is then cached (didn't look into how that works exactly). So if I just write the new data into the global keyset instead, everything should still work.

Basically, I would replace the `for` loop in this function:
https://github.com/ElektraInitiative/libelektra/blob/a5d9a51c0366d7dc8ed70a7808d1daf113c62b30/src/libs/elektra/split.c#L828-L938

Reply:
I think this should work. Note that I also needed to store the old parentKey name etc., but basically that's all there is to it.